daemon, k8s: Convert CEP event handling to subscriber model

christarazi opened the pull request:
As part of https://github.com/cilium/cilium/issues/15471, move the CEP
handling to the new subscriber model.

This commit should have no functional changes.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/k8s/cep/cep.go:
```diff
@@ -0,0 +1,168 @@
+// Copyright Authors of Cilium
+
+// Package cep contains the CiliumEndpoint event handling logic from Kubernetes.
+package cep
```
Open to feedback on whether this logic is in the right location. Another option would be another subpackage under `pkg/ipcache` since most of this logic is about ipcache handling with CEPs.

christarazi commented on pkg/k8s/cep/cep.go:
```diff
@@ -0,0 +1,168 @@
+	return s.OnUpdateCiliumEndpoint(nil, ep)
+}
+
+func (s *Subscriber) OnUpdateCiliumEndpoint(oldCEP, newCEP *types.CiliumEndpoint) error {
```
This logic is almost pure copy-paste from k8s watcher, for which the main motivation is to clean up the k8s watcher and slowly remove business logic from it.

aanm reviewed (commented):
Did we discuss the concern of having one "subscriber" slowing down all of the remaining subscribers?

I understand that the current model, the "non-subscriber" model, is that one feature can slow down the remaining feature because all of the code is sequencial. However, if we move to a model where we have one queue for each k8s type per feature, or per subscriber, it would result in an independent realization state for each subscriber. This essentially means that instead of having a list of subscribers to go over each time we receive an event we would have a list of queues to go over each time we receive an event. The queues are non-blocking where the subscriber are blocking.

christarazi committed:
daemon, k8s: Convert CEP event handling to subscriber model

As part of https://github.com/cilium/cilium/issues/15471, move the CEP
handling to the new subscriber model.

This commit should have no functional changes.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
